src/allmydata/test/test_istorageserver.py

```diff
@@ -0,0 +1,282 @@
+    )
+
+    @inlineCallbacks
+    def test_allocate_buckets_more_sharenums(self):
+        """
+        allocate_buckets() with the same storage index but more sharenums
+        acknowledges the extra shares don't exist.
+
+        Fails due to https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3793
+        """
+        si, renew_secret, cancel_secret = (
+            new_storage_index(),
+            new_secret(),
+            new_secret(),
+        )
+        yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+        (already_got2, allocated2) = yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(7)),
+            1024,
+            Referenceable(),
+        )
+        self.assertEqual(already_got2, set())  # none were fully written
+        self.assertEqual(set(allocated2.keys()), set(range(7)))
+
+    test_allocate_buckets_more_sharenums.skip = (
+        "https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3793"
+    )
+
+    @inlineCallbacks
+    def test_written_shares_are_allocated(self):
+        """
+        Shares that are fully written to show up as allocated in result from
+        ``IStoragServer.allocate_buckets()``.  Partially-written or empty
+        shares don't.
+        """
+        si, renew_secret, cancel_secret = (
+            new_storage_index(),
+            new_secret(),
+            new_secret(),
+        )
+        (_, allocated) = yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+
+        # Bucket 1 is fully written in one go.
+        yield allocated[1].callRemote("write", 0, b"1" * 1024)
+        yield allocated[1].callRemote("close")
+
+        # Bucket 2 is fully written in two steps.
+        yield allocated[2].callRemote("write", 0, b"1" * 512)
+        yield allocated[2].callRemote("write", 512, b"2" * 512)
+        yield allocated[2].callRemote("close")
+
+        # Bucket 0 has partial write.
+        yield allocated[0].callRemote("write", 0, b"1" * 512)
+
+        (already_got, _) = yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+        self.assertEqual(already_got, {1, 2})
+
+    @inlineCallbacks
+    def test_written_shares_are_readable(self):
+        """
+        Shares that are fully written to can be read.
+
+            1. The result is not affected by the order in which writes
+               happened, only by their offsets.
+
+            2. When overlapping writes happen, the resulting read returns the
+               earliest written value.
```

Comment:
GBS explicitly forbids overlapping writes (the `CONFLICT` status case of https://tahoe-lafs.readthedocs.io/en/latest/proposed/http-storage-node-protocol.html#put-v1-immutable-storage-index-share-number).

I'm somewhat dismayed that StorageServer allows it ... I'm not sure if the fact that it *silently discards* the newer data makes it better or worse (but "silently does something surprising" doesn't *surprise* me).  

I wonder if clients do this and if so, why.

Reply:
That docstring is wrong, it actually returns the latest written value.

Reply:
Should I not be testing overlapping writes succeed, then?

Reply:
The current behavior seems like a bad day waiting to happen.  I don't want to constrain GBS to that behavior.  So I guess either we don't include this behavior in the interface test or we ... change the current behavior?

Here's what I can imagine for the current behavior:

* No overlapping writes happen.  Everything is fine and also we don't have to allow overlapping writes for this to work.
* Overlapping writes happen and the data is the same each time.  Everything is fine but only because we allow the overlap.  It doesn't matter if we keep old or new because it's the same.  I suspect we can't signal an error here because clients won't know to handle it and will fail out of their upload attempt.
* Overlapping writes happen and the data is different each time.  This is terri-bad and it would be better if the server blew up the client attempting the upload.  That doesn't happen with the current implementation.

I can see a path where we:

1. change the current behavior so that if there is overlap and the data differs then it's a hard failure.
2. change GBS spec to allow overlaps as long as the data is the same.

I *am* slightly afraid that somehow the current client depends on last-write-wins with non-matching data behavior and we'll break real uploads if we break this case. :(  If that's the case, it is also totally horrible but it still means we can't just forbid that behavior right now. :(

I'm also slightly afraid of how much complexity is involved in detecting that overlapped writes contain matching data.  It changes a fairly simple "write this sequence of bytes to a file" to something with more state and logic. :(

I'm also not thrilled about GBS being forced to take on that same complexity to match the proposed behavior. :(

So I think this plan sucks.  Should I just get over it and accept that GBS will have to accept overlapped writes with non-matching data too?

Reply:
These are immutables? So having non-matching data seems like an integrity bug worth fixing.

As far as complexity: given parallelism inherent in HTTP, there is no guarantee that writes will arrive in order, unlike current Foolscap. So there's already more complexity due to holes that in practice probably don't happen much with Foolscap. So while adding checks adds complexity, it also is probably a good safety measure. And it seems quite amenable to property-based testing, for example.

So I think we should just do it.

Reply:
Discussion moved to https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3801